File: ai-solutions/05/gpt-5.1-codex/python/main.py
from __future__ import annotations

from bisect import bisect_right
from pathlib import Path
from typing import Iterable, List, Sequence, Tuple


Range = Tuple[int, int]


def read_input(filepath: Path) -> Tuple[List[Range], List[int]]:
    """Parse the puzzle input into fresh ranges and ingredient ids."""
    content = filepath.read_text().strip().split("\n\n")
    if len(content) != 2:
        raise ValueError("Input file must contain ranges, a blank line, then IDs.")

    raw_ranges, raw_ids = content
    ranges: List[Range] = []
    for line in raw_ranges.strip().splitlines():
        start, end = line.split("-")
        ranges.append((int(start), int(end)))

    ids = [int(line) for line in raw_ids.strip().splitlines() if line]
    return ranges, ids


def merge_ranges(ranges: Iterable[Range]) -> List[Range]:
    """Merge overlapping or adjacent ranges to simplify later lookups."""
    sorted_ranges = sorted(ranges)
    if not sorted_ranges:
        return []

    merged: List[Range] = []
    cur_start, cur_end = sorted_ranges[0]

    for start, end in sorted_ranges[1:]:
        if start <= cur_end + 1:
            cur_end = max(cur_end, end)
        else:
            merged.append((cur_start, cur_end))
            cur_start, cur_end = start, end

    merged.append((cur_start, cur_end))
    return merged


def count_fresh(ids: Sequence[int], merged_ranges: Sequence[Range]) -> int:
    """Count how many ids fall inside any merged range using binary search."""
    if not merged_ranges:
        return 0

    starts = [r[0] for r in merged_ranges]
    fresh = 0
    for value in ids:
        idx = bisect_right(starts, value) - 1
        if idx >= 0 and value <= merged_ranges[idx][1]:
            fresh += 1
    return fresh


def total_fresh_ids(merged_ranges: Sequence[Range]) -> int:
    return sum(end - start + 1 for start, end in merged_ranges)


def locate_input() -> Path:
    script_dir = Path(__file__).resolve().parent
    project_root = script_dir
    for _ in range(4):
        project_root = project_root.parent

    candidates = [
        project_root / "inputs" / "05.txt",
        Path("inputs/05.txt"),
    ]

    for candidate in candidates:
        if candidate.exists():
            return candidate

    raise FileNotFoundError("Unable to locate inputs/05.txt")


def main() -> None:
    input_path = locate_input()
    ranges, ids = read_input(input_path)
    merged = merge_ranges(ranges)

    part1 = count_fresh(ids, merged)
    part2 = total_fresh_ids(merged)

    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()

File: ai-solutions/05/gpt-5.1-codex/rust/src/main.rs
use std::cmp;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug)]
struct Range {
    start: u64,
    end: u64,
}

fn parse_input(path: &Path) -> (Vec<Range>, Vec<u64>) {
    let content = fs::read_to_string(path).expect("failed to read input file");
    let mut sections = content.trim().split("\n\n");
    let ranges_section = sections
        .next()
        .expect("input must contain a ranges section");
    let ids_section = sections
        .next()
        .expect("input must contain an ids section");

    let ranges = ranges_section
        .lines()
        .map(|line| {
            let mut parts = line.split('-');
            let start: u64 = parts
                .next()
                .expect("missing range start")
                .parse()
                .expect("invalid range start");
            let end: u64 = parts
                .next()
                .expect("missing range end")
                .parse()
                .expect("invalid range end");
            Range { start, end }
        })
        .collect::<Vec<_>>();

    let ids = ids_section
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| line.parse::<u64>().expect("invalid id"))
        .collect::<Vec<_>>();

    (ranges, ids)
}

fn merge_ranges(ranges: &[Range]) -> Vec<Range> {
    if ranges.is_empty() {
        return Vec::new();
    }

    let mut sorted = ranges.to_vec();
    sorted.sort_by_key(|r| r.start);

    let mut merged = Vec::with_capacity(sorted.len());
    let mut current = sorted[0];

    for range in sorted.into_iter().skip(1) {
        if range.start <= current.end + 1 {
            current.end = cmp::max(current.end, range.end);
        } else {
            merged.push(current);
            current = range;
        }
    }

    merged.push(current);
    merged
}

fn is_fresh(id: u64, merged: &[Range]) -> bool {
    if merged.is_empty() {
        return false;
    }

    let mut left = 0;
    let mut right = merged.len();

    while left < right {
        let mid = left + (right - left) / 2;
        let range = merged[mid];

        if id < range.start {
            right = mid;
        } else if id > range.end {
            left = mid + 1;
        } else {
            return true;
        }
    }

    false
}

fn count_fresh(ids: &[u64], merged: &[Range]) -> usize {
    ids.iter().filter(|&&id| is_fresh(id, merged)).count()
}

fn total_fresh(merged: &[Range]) -> u64 {
    merged.iter().map(|r| r.end - r.start + 1).sum()
}

fn locate_input() -> PathBuf {
    let manifest_candidate =
        Path::new(env!("CARGO_MANIFEST_DIR")).join("../../../../inputs/05.txt");
    if manifest_candidate.exists() {
        return manifest_candidate;
    }

    let cwd_candidate = Path::new("inputs/05.txt");
    if cwd_candidate.exists() {
        return cwd_candidate.to_path_buf();
    }

    panic!("Unable to locate inputs/05.txt");
}

fn main() {
    let input_path = locate_input();
    let (ranges, ids) = parse_input(&input_path);
    let merged = merge_ranges(&ranges);

    let part1 = count_fresh(&ids, &merged);
    let part2 = total_fresh(&merged);

    println!("{}", part1);
    println!("{}", part2);
}


